px.py

```diff
@@ -552,26 +575,31 @@ def do_proxy_type(self):
                 return 408, None, None
 
         # New proxy, don't know type yet
-        if self.proxy_address not in State.proxy_type:
-            dprint("Searching proxy type")
-            resp, headers, body = self.do_socket()
+        # Do locking to avoid updating State.proxy_type simultaneously by multiple threads
+        State.proxy_type_lock.acquire()
+        try:
+            if self.proxy_address not in State.proxy_type:
+                dprint("Searching proxy type")
+                resp, headers, body = self.do_socket()
 
-            proxy_auth = ""
-            for header in headers:
-                if header[0] == "Proxy-Authenticate":
-                    proxy_auth += header[1] + " "
+                proxy_auth = ""
+                for header in headers:
+                    if header[0] == "Proxy-Authenticate":
+                        proxy_auth += header[1] + " "
 
-            if "NTLM" in proxy_auth.upper():
-                State.proxy_type[self.proxy_address] = "NTLM"
-            elif "KERBEROS" in proxy_auth.upper():
-                State.proxy_type[self.proxy_address] = "KERBEROS"
-            elif "NEGOTIATE" in proxy_auth.upper():
-                State.proxy_type[self.proxy_address] = "NEGOTIATE"
+                if "NTLM" in proxy_auth.upper():
+                    State.proxy_type[self.proxy_address] = "NTLM"
+                elif "KERBEROS" in proxy_auth.upper():
+                    State.proxy_type[self.proxy_address] = "KERBEROS"
+                elif "NEGOTIATE" in proxy_auth.upper():
+                    State.proxy_type[self.proxy_address] = "NEGOTIATE"
 
-            dprint("Auth mechanisms: " + proxy_auth)
-            dprint("Selected: " + str(State.proxy_type))
+                dprint("Auth mechanisms: " + proxy_auth)
+                dprint("Selected: " + str(State.proxy_type))
 
```

Comment:
Does the entire block need to be locked or just the portion where it's being written to?

Reply:
If only the write part is locked two threads could try to get type of the same proxy in parallel. As long as the proxy's answers are consistent it shouldn't do any harm. 
Hm, it could be indeed better to lock only the write part so that two threads asking different proxies can work in parallel.

Reply:
I saw this in the code where we are locking proxy_mode and proxy_type as well, large chunk. Hoping we could minimize the size of the section there as well.

Reply:
I removed the locking for proxy_type in load_proxy now as it is only a simple assignment. But the locking in do_proxy_type remains. I encountered that loading a "modern" web page with many images results in many parallel requests. Without locking px tries to get the proxy type for many of these requests in parallel too. Getting the proxy type from a proxy with auth needs an additional request/response roundtrip over the network. Even if the connection will be reused I assume that transfering data over the network is slower then locking.
Looking at this I found another race condition. If one thread executed do_proxy_type, then another thread calls load_proxy which clears State.proxy_type and then the first thread reads the proxy type from State_proxy it would get None. To avoid this I changed do_proxy_type to return the proxy_type as part of the function result und to use State.proxy_type in a thread safe manner (one read, one write). State.proxy_type is now filled and queried only in do_proxy_type and cleared in load_proxy.